### PR TITLE
Add support for Incoming Webhooks

### LINKED ISSUE #197: Support for incoming webhooks?
Are y'all interested in exposing an object for posting [incoming webhooks](https://api.slack.com/incoming-webhooks)?

I was thinking an interface like this:

```js
var IncomingWebhook = require('@slack/client').IncomingWebhook;

// slackUrl is the webhook url provided when creating a new incoming webhook
// options would mirror the payload, so you could set default bot name, bot channel, bot emoji, etc
var iw = new IncomingWebhook(slackUrl, options);

// can pass a string argument
iw.send('some text', cb); // sends a payload of { "text": "some text" } along with any defaults set up in the initialization options

// can pass a payload object, anything not provided uses initialization defaults
iw.send({
  text: 'some text',
  attachments: [
    // attachment data
  ], 
  username: 'some username',
  icon_emoji: ':robot_face:'
}, cb);
```

If that's desired, I'm interested in working on it.

### PR DESCRIPTION
**Pull request by crookedneighbor**
Closes #197 

Tried to match the code style of the rest of the project, but if I didn't do something right, feel free to tell me. Happy to adjust.

### CODE REVIEW

File: lib/clients/transports/request.js [hunk 1]
```diff
@@ -34,6 +34,10 @@ var getRequestTransportArgs = function getReqestTransportArgs(args) {
     transportArgs.form = args.data;
   }
 
+  if (has(args, 'body')) {
```
Comment (crookedneighbor) [inline lib/clients/transports/request.js]: There was no test coverage around this file that I could fine. Is it cool to add this?

File: lib/clients/incoming-webhook/client.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  this._slackUrl = slackUrl;
+  this.defaults = {
+    username: _defaults.username,
+    iconEmoji: _defaults.iconEmoji,
+    channel: _defaults.channel,
+    text: _defaults.text
+  };
+
+  this._transport = _defaults._transport || requestsTransport;
```
Comment (crookedneighbor) [inline lib/clients/incoming-webhook/client.js]: I had this be injectable for easy stubbing. Could also accomplish the same thing by stubbing the transportRequests object directly. This was the pattern I saw elsewhere, but am happy to change it.

File: lib/clients/incoming-webhook/client.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+/**
+ * Formats the data to be delivered to the transport
+ * @private
+ */
+IncomingWebhook.prototype._formatData = function _formatData(message) {
+  var data = {
+    username: this.defaults.username,
+    icon_emoji: this.defaults.iconEmoji,
```
Comment (crookedneighbor) [inline lib/clients/incoming-webhook/client.js]: I figured you wanted the properties to be in camelcase, and have the module convert this one to snakecase.

Comment (crookedneighbor) [inline lib/clients/incoming-webhook/client.js]: Also, should this do some convenience checking so you can pass with or without `:` and have it work? Ie, make `robot_face` and `:robot_face:` both valid values?

File: lib/clients/incoming-webhook/client.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * @param {string} defaults.channel The default channel to use when sending a webhook.
+ *      If no channel is specified, the one chosen when creating the webhook will be used.
+ * @param {string} defaults.text The default text to use when sending a webhook.
+ * @constructor
+ */
+function IncomingWebhook(slackUrl, defaults) {
+  var _defaults = defaults || {};
+
+  if (!slackUrl) {
+    throw new Error('Slack url is required');
```
Comment (crookedneighbor) [inline lib/clients/incoming-webhook/client.js]: I saw in some other places, there were things like `RTMSlackError`. Should this be a new kind of Slack Error? Or does it matter?

File: lib/clients/incoming-webhook/client.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+
+/**
+ * Sends a message via an incoming webhook
+ * @param {String|Object} message The message to send. Can be text or an object that
+ *     overrides the defaults in initialization.
+ * @param {Function=} optCb
+ */
+IncomingWebhook.prototype.send = function send(message, cb) {
```
Comment (crookedneighbor) [inline lib/clients/incoming-webhook/client.js]: Should this return a Promise if a callback isn't provided? Or just leave that up to the user to wrap it if they prefer promises?

**Comment by l12s** [comment]
I've been on vacation, so sorry about the slow response. Will check it out properly later this week - thanks!

**Comment by crookedneighbor** [comment]
:+1:

**Comment by crookedneighbor** [comment]
Any update on this?

**Comment by DEGoodmanWilson** [comment]
Hello! My apologies for not commenting sooner. I think this is an interesting and valuable feature, but I first need to get through a backlog of more pressing tasks. I promise that I am going to come around and have a look at this PR in the coming weeks, and I want to give this PR the attention it deserves. But I've got to find the time to do that first ;)

**Comment by crookedneighbor** [comment]
No worries, I understand how that goes. Happy to help any way I can.

**Comment by crookedneighbor** [comment]
Just checking on the status of this. Anything I can do to help?

File: lib/clients/incoming-webhook/client.js (hunk 4 above)
Comment (DEGoodmanWilson) [inline lib/clients/incoming-webhook/client.js]: yeah, each of the clients has its own kind of error defined in errors.js. Let's add one for incoming webhooks. But the thing is, we don't throw errors, rather we either call error handling callbacks, or emit error events as the case may be. Of course, we aren't doing checking of required parameters in the other client constructors, so maybe this is a fine thing to do, actually. Let's leave this as you have it until we have a good reason to do this differently.

File: lib/clients/incoming-webhook/client.js (hunk 5 above)
Comment (DEGoodmanWilson) [inline lib/clients/incoming-webhook/client.js]: I think this is fine.

File: lib/clients/incoming-webhook/client.js (hunk 3 above)
Comment (DEGoodmanWilson) [inline lib/clients/incoming-webhook/client.js]: I would be of the mind to require the `:emoji:` form, and then do some checking to make sure what we got passed is of that form. We can add that later ;)

File: lib/clients/transports/request.js (hunk 1 above)
Comment (DEGoodmanWilson) [inline lib/clients/transports/request.js]: Yeah, this is fine. We should write some tests around this whole file tho

**Comment by DEGoodmanWilson** [comment]
No worries. Finally had a chance to look. Thanks for pinging me! I'm going to approve this, looks good.

**Comment by crookedneighbor** [comment]
:+1:

**Comment by crookedneighbor** [comment]
Any idea when you'll publish this to NPM?

**Pull request merged**